platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/ReservedSignedState.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.platform.state.signed;
+
+import static com.swirlds.logging.LogMarker.EXCEPTION;
+
+import com.swirlds.common.AutoCloseableNonThrowing;
+import com.swirlds.common.exceptions.ReferenceCountException;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicLong;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * A wrapper around a signed state that holds a reservation. Until this wrapper is released/closed, the signed state
+ * contained within will not be destroyed.
+ * <p>
+ * This class is not thread safe. That is, it is not safe for one thread to access this object while another thread is
+ * asynchronously closing it. Each thread should hold its own reservation on a state if it needs to access a state.
+ */
+public class ReservedSignedState implements AutoCloseableNonThrowing {
+
+    private static final Logger logger = LogManager.getLogger(ReservedSignedState.class);
+
+    private final SignedState signedState;
+    private final String reason;
+    private final long reservationId = nextReservationId.getAndIncrement();
+    private boolean closed = false;
+
+    /**
+     * The next reservation id to use. It is ok that this is static, since we don't care which ID any particular
+     * reservation has, as long as that ID is unique.
+     */
+    private static final AtomicLong nextReservationId = new AtomicLong(0);
+
+    /**
+     * Create a wrapper around null (for scenarios where we are storing a null signed state).
+     */
+    public ReservedSignedState() {
```

Comment:
Can we make that constructor non-public? Does it make sense to define the wrapper around null as a singleton?

Reply:
I'm not sure it should be a singleton, since it would be preferable to throw an exception if the object is closed more than once.

I've made this constructor non-public, and instead added a method `static @NonNull ReservedSignedState createNullReservation()`